ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/Diff.hs

```diff
@@ -0,0 +1,227 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs #-}
+
+module Ouroboros.Consensus.Ledger.Tables.Diff (
+    -- * Types
+    Delta (..)
+  , Diff (..)
+    -- * Conversion
+  , keysSet
+    -- * Construction
+  , diff
+    -- ** Maps
+  , fromMap
+  , fromMapDeletes
+  , fromMapInserts
+    -- ** Set
+  , fromSetDeletes
+    -- ** Lists
+  , fromList
+  , fromListDeletes
+  , fromListInserts
+    -- * Query
+    -- ** Size
+  , null
+  , numDeletes
+  , numInserts
+  , size
+    -- * Applying diffs
+  , applyDiff
+  , applyDiffForKeys
+    -- * Filter
+  , filterOnlyKey
+  , foldMapDelta
+  , fromAntiDiff
+  , toAntiDiff
+  , traverseDeltaWithKey_
+  ) where
+
+import           Control.Monad (void)
+import           Data.Bifunctor
+import           Data.Foldable (foldMap')
+import qualified Data.Map.Diff.Strict.Internal as Anti
+import qualified Data.Map.Merge.Strict as Merge
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Monoid
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           GHC.Generics
+import           NoThunks.Class
+import           Prelude hiding (null)
+
+{------------------------------------------------------------------------------
+  Types
+------------------------------------------------------------------------------}
+
+newtype Diff k v = Diff (Map k (Delta v))
+  deriving stock (Show, Eq)
+  deriving Generic
+  deriving newtype NoThunks
+
+-- | Custom 'Functor' instance, since @'Functor' ('Map' k)@ is actually the
+-- 'Functor' instance for a lazy Map.
+instance Functor (Diff k) where
+  fmap f (Diff m) = Diff $ Map.map (fmap f) m
+
+instance Ord k => Semigroup (Diff k v) where
+  (<>) :: Diff k v -> Diff k v -> Diff k v
+  (Diff m1) <> (Diff m2) = Diff $ Map.unionWith (<>) m1 m2
+
+instance Ord k => Monoid (Diff k v) where
+  mempty :: Diff k v
+  mempty = Diff mempty
+
+data Delta v =
+    Insert !v
+  | Delete
+  deriving stock (Show, Eq, Functor)
+  deriving Generic
+  deriving NoThunks
+
+-- | Right-biased
```

Comment:
The case you're describing is 

```haskell
Insert x <> Delete = Delete
```

I've considered whether we could in that case remove the key from the `Diff`,  but we'll have to keep around the `Delete`, because

```haskell
  applyDiff (Map.fromList [(1, 'a')]) (Diff (Map.fromList [(1, Delete)]))
/= 
  applyDiff (Map.fromList [(1, 'a')]) (Diff Map.empty)
```